File: 프로그래머스/숫자 변환하기.ts
/*
https://school.programmers.co.kr/learn/courses/30/lessons/154538

1. 문제분석
자연수 x를 y로 변환 -> x+n, x*2, x*3만 가능
 -> 필요한 최소 연산횟수 리턴 -> 불가능하면 -1
x, y는 1~100만, n은 1~(100만-1)

2. 풀어보기
x=10,y=40,n=5 -> 10*2*2 -> 2
x=10,y=40,n=30 -> 10+30 -> 1
x=2,y=5,n=4 -> 불가 -> -1

3. 슈도코드
연산 x+n, x*2, x*3 정의 -> BFS로 돌리면서, 연산회수 세다가
 -> y와 같아지면 연산회수 리턴 -> 다 넘쳐버리면 -1 리턴

4. 구현코드

5. 풀이회고
DP로 리팩토링
*/
{
const solution = (x: number, y: number, n: number): number => {
  if (x === y) {
    return 0;
  }

  return by_bfs(x, y, n);
};

const by_bfs = (x: number, y: number, n: number) => {
  const queue: [number, number][] = [[x, 0]];
  let first = 0;

  while (first < queue.length) {
    const [sum, count] = queue[first++];
    if (sum === y) {
      return count;
    }

    if (sum < y) {
      queue.push([sum + n, count + 1]);
      queue.push([sum * 2, count + 1]);
      queue.push([sum * 3, count + 1]);
    }
  }

  return -1;
};

console.log(solution(10, 40, 5)); // 2
console.log(solution(10, 40, 30)); // 1
console.log(solution(2, 5, 4)); // -1
}

File: 프로그래머스/lv2/지준혁/숫자변환하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>

using namespace std;

// https://school.programmers.co.kr/learn/courses/30/lessons/154538

// 2를 곱하거나 3을 곱하거나 N을 더할 때 최소 연산 횟수 => bfs로 접근
// 1. 종료조건: x가 y와 같을 때, 같은 경우를 발견하지 못할 때 -1
// 2. 큐에다가 가능한 수 넣다뺐다하면서 탐색
// 3. 1'000'000이 넘어간 수는 더 이상 넣지 않음(탐색 안함)

bool is_visited[1'000'004];

int bfs(int x, int y, int n) {
    queue<pair<int, int>> q;

    q.push({x, 0});
    is_visited[x] = true;

    while (!q.empty()) {
        auto [cur_x, cnt] = q.front();
        q.pop();

        if (cur_x == y) {
            return cnt;
        }

        if (cur_x > 1'000'000) {
            continue;
        }

        if (cur_x * 2 <= 1'000'000 && !is_visited[cur_x * 2]) {
            is_visited[cur_x * 2] = true;
            q.push({cur_x * 2, cnt + 1});
        }
        if (cur_x * 3 <= 1'000'000 && !is_visited[cur_x * 3]) {
            is_visited[cur_x * 3] = true;
            q.push({cur_x * 3, cnt + 1});
        }
        if (cur_x + n <= 1'000'000 && !is_visited[cur_x + n]) {
            is_visited[cur_x + n] = true;
            q.push({cur_x + n, cnt + 1});
        }
    }
    return -1;
}

int solution(int x, int y, int n) {
    int answer = 0;
        
    answer = bfs(x, y, n);;
    
    return answer;
}

